// @trapezegroup/core-webui-styles/css-variables/popup.css.js
export const style = `/*!
 * # Fomantic-UI - Popup
 * http://github.com/fomantic/Fomantic-UI/
 *
 *
 * Released under the MIT license
 * http://opensource.org/licenses/MIT
 *
 */


/*******************************
            Popup
*******************************/

:host(.ui.popup),
.ui.popup {
  display: none;
  position: absolute;
  top: 0;
  right: 0;
  
/* Fixes content being squished when inline (moz only) */
  min-width: -webkit-min-content;
  min-width: -moz-min-content;
  min-width: min-content;
  z-index: 1900;
  border: 1px solid var(--prim-brd);
  line-height: 1.4285em;
  max-width: 250px;
  background: var(--prim-bg);
  padding: 0.833em 1em;
  font-weight: normal;
  font-style: normal;
  color: var(--prim-txt);
  border-radius: 0.33333333rem;
  box-shadow: var(--prim-shdw);
}
:host(.ui.popup) > .header,
.ui.popup > .header {
  padding: 0;
  font-family: 'Source Sans Pro', 'Segoe UI', 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-size: 1.16666667em;
  line-height: 1.2;
  font-weight: bold;
}
:host(.ui.popup) > .header + .content,
.ui.popup > .header + .content{
  padding-top: 0.58333333em;
}
:host(.ui.popup):before,
.ui.popup:before {
  position: absolute;
  content: '';
  width: 0.83333333em;
  height: 0.83333333em;
  background: var(--prim-bg);
  transform: rotate(45deg);
  z-index: 1901;
  box-shadow: 1px 1px 0 0 var(--prim-bg);
}


/*******************************
            Types
*******************************/


/*--------------
    Tooltip
---------------*/


/* Content */
[data-tooltip] {
  position: relative;
}

/* Arrow */
[data-tooltip]:before {
  pointer-events: none;
  position: absolute;
  content: '';
  font-size: 1rem;
  width: 0.83333333em;
  height: 0.83333333em;
  background: var(--prim-bg);
  -webkit-transform: rotate(45deg);
          transform: rotate(45deg);
  z-index: 1901;
  -webkit-box-shadow: 1px 1px 0 0 var(--prim-bg);
          box-shadow: 1px 1px 0 0 var(--prim-bg);
}

/* Popup */
[data-tooltip]:after {
  pointer-events: none;
  content: attr(data-tooltip);
  position: absolute;
  text-transform: none;
  text-align: left;
  white-space: nowrap;
  font-size: 1rem;
  border: 1px solid var(--prim-brd);
  line-height: 1.4285em;
  max-width: none;
  background: var(--prim-bg);
  padding: 0.833em 1em;
  font-weight: normal;
  font-style: normal;
  color: var(--prim-txt);
  border-radius: 0.33333333rem;
  box-shadow: var(--prim-shdw);
  z-index: 1900;
}

/* Default Position (Top Center) */
[data-tooltip]:not([data-position]):before {
  top: auto;
  right: auto;
  bottom: 100%;
  left: 50%;
  background: var(--prim-bg);
  margin-left: -0.08333333rem;
  margin-bottom: 0.16666667rem;
}
[data-tooltip]:not([data-position]):after {
  left: 50%;
  -webkit-transform: translateX(-50%);
          transform: translateX(-50%);
  bottom: 100%;
  margin-bottom: 0.58333333em;
}

/* Animation */
[data-tooltip]:before,
[data-tooltip]:after {
  pointer-events: none;
  visibility: hidden;
}
[data-tooltip]:before {
  opacity: 0;
  -webkit-transform: rotate(45deg) scale(0) !important;
          transform: rotate(45deg) scale(0) !important;
  -webkit-transform-origin: center top;
          transform-origin: center top;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
}
[data-tooltip]:after {
  opacity: 1;
  -webkit-transform-origin: center bottom;
          transform-origin: center bottom;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
}
[data-tooltip]:hover:before,
[data-tooltip]:hover:after {
  visibility: visible;
  pointer-events: auto;
}
[data-tooltip]:hover:before {
  -webkit-transform: rotate(45deg) scale(1) !important;
          transform: rotate(45deg) scale(1) !important;
  opacity: 1;
}

/* Animation Position */
[data-tooltip]:after,
[data-tooltip][data-position="top center"]:after,
[data-tooltip][data-position="bottom center"]:after {
  -webkit-transform: translateX(-50%) scale(0) !important;
          transform: translateX(-50%) scale(0) !important;
}
[data-tooltip]:hover:after,
[data-tooltip][data-position="bottom center"]:hover:after {
  -webkit-transform: translateX(-50%) scale(1) !important;
          transform: translateX(-50%) scale(1) !important;
}
[data-tooltip][data-position="left center"]:after,
[data-tooltip][data-position="right center"]:after {
  -webkit-transform: translateY(-50%) scale(0) !important;
          transform: translateY(-50%) scale(0) !important;
}
[data-tooltip][data-position="left center"]:hover:after,
[data-tooltip][data-position="right center"]:hover:after {
  -webkit-transform: translateY(-50%) scale(1) !important;
          transform: translateY(-50%) scale(1) !important;
}
[data-tooltip][data-position="top left"]:after,
[data-tooltip][data-position="top right"]:after,
[data-tooltip][data-position="bottom left"]:after,
[data-tooltip][data-position="bottom right"]:after {
  -webkit-transform: scale(0) !important;
          transform: scale(0) !important;
}
[data-tooltip][data-position="top left"]:hover:after,
[data-tooltip][data-position="top right"]:hover:after,
[data-tooltip][data-position="bottom left"]:hover:after,
[data-tooltip][data-position="bottom right"]:hover:after {
  -webkit-transform: scale(1) !important;
          transform: scale(1) !important;
}

/*--------------
    Inverted
---------------*/


/* Arrow */
[data-tooltip][data-inverted]:before {
  -webkit-box-shadow: none !important;
          box-shadow: none !important;
}

/* Arrow Position */
[data-tooltip][data-inverted]:before {
  background: #1D252D;
}

/* Popup  */
[data-tooltip][data-inverted]:after {
  background: #1D252D;
  color: var(--prim-bg);
  border: none;
  -webkit-box-shadow: none;
          box-shadow: none;
}
[data-tooltip][data-inverted]:after .header {
  background: none;
  color: var(--prim-bg);
}

/*--------------
    Position
---------------*/


/* Top Center */
[data-position="top center"][data-tooltip]:after {
  top: auto;
  right: auto;
  left: 50%;
  bottom: 100%;
  -webkit-transform: translateX(-50%);
          transform: translateX(-50%);
  margin-bottom: 0.58333333em;
}
[data-position="top center"][data-tooltip]:before {
  top: auto;
  right: auto;
  bottom: 100%;
  left: 50%;
  background: var(--prim-bg);
  margin-left: -0.08333333rem;
  margin-bottom: 0.16666667rem;
}

/* Top Left */
[data-position="top left"][data-tooltip]:after {
  top: auto;
  right: auto;
  left: 0;
  bottom: 100%;
  margin-bottom: 0.58333333em;
}
[data-position="top left"][data-tooltip]:before {
  top: auto;
  right: auto;
  bottom: 100%;
  left: 1em;
  margin-left: -0.08333333rem;
  margin-bottom: 0.16666667rem;
}

/* Top Right */
[data-position="top right"][data-tooltip]:after {
  top: auto;
  left: auto;
  right: 0;
  bottom: 100%;
  margin-bottom: 0.58333333em;
}
[data-position="top right"][data-tooltip]:before {
  top: auto;
  left: auto;
  bottom: 100%;
  right: 1em;
  margin-left: -0.08333333rem;
  margin-bottom: 0.16666667rem;
}

/* Bottom Center */
[data-position="bottom center"][data-tooltip]:after {
  bottom: auto;
  right: auto;
  left: 50%;
  top: 100%;
  -webkit-transform: translateX(-50%);
          transform: translateX(-50%);
  margin-top: 0.58333333em;
}
[data-position="bottom center"][data-tooltip]:before {
  bottom: auto;
  right: auto;
  top: 100%;
  left: 50%;
  margin-left: -0.08333333rem;
  margin-top: 0.16666667rem;
}

/* Bottom Left */
[data-position="bottom left"][data-tooltip]:after {
  left: 0;
  top: 100%;
  margin-top: 0.58333333em;
}
[data-position="bottom left"][data-tooltip]:before {
  bottom: auto;
  right: auto;
  top: 100%;
  left: 1em;
  margin-left: -0.08333333rem;
  margin-top: 0.16666667rem;
}

/* Bottom Right */
[data-position="bottom right"][data-tooltip]:after {
  right: 0;
  top: 100%;
  margin-top: 0.58333333em;
}
[data-position="bottom right"][data-tooltip]:before {
  bottom: auto;
  left: auto;
  top: 100%;
  right: 1em;
  margin-left: -0.16666667rem;
  margin-top: 0.08333333rem;
}

/* Left Center */
[data-position="left center"][data-tooltip]:after {
  right: 100%;
  top: 50%;
  margin-right: 0.58333333em;
  -webkit-transform: translateY(-50%);
          transform: translateY(-50%);
}
[data-position="left center"][data-tooltip]:before {
  right: 100%;
  top: 50%;
  margin-top: -0.16666667rem;
  margin-right: -0.08333333rem;
}

/* Right Center */
[data-position="right center"][data-tooltip]:after {
  left: 100%;
  top: 50%;
  margin-left: 0.58333333em;
  -webkit-transform: translateY(-50%);
          transform: translateY(-50%);
}
[data-position="right center"][data-tooltip]:before {
  left: 100%;
  top: 50%;
  margin-top: -0.08333333rem;
  margin-left: 0.16666667rem;
}

/* Arrow */
[data-position~="bottom"][data-tooltip]:before {
  background: var(--prim-bg);
  box-shadow: -1px -1px 0 0 var(--prim-bg);
}
[data-position="left center"][data-tooltip]:before {
  background: var(--prim-bg);
  box-shadow: 1px -1px 0 0 var(--prim-bg);
}
[data-position="right center"][data-tooltip]:before {
  background: var(--prim-bg);
  box-shadow: -1px 1px 0 0 var(--prim-bg);
}
[data-position~="top"][data-tooltip]:before {
  background: var(--prim-bg);
}

/* Inverted Arrow Color */
[data-inverted][data-position~="bottom"][data-tooltip]:before {
  background: #1D252D;
  -webkit-box-shadow: -1px -1px 0 0 var(--prim-bg);
          box-shadow: -1px -1px 0 0 var(--prim-bg);
}
[data-inverted][data-position="left center"][data-tooltip]:before {
  background: #1D252D;
  -webkit-box-shadow: 1px -1px 0 0 var(--prim-bg);
          box-shadow: 1px -1px 0 0 var(--prim-bg);
}
[data-inverted][data-position="right center"][data-tooltip]:before {
  background: #1D252D;
  -webkit-box-shadow: -1px 1px 0 0 var(--prim-bg);
          box-shadow: -1px 1px 0 0 var(--prim-bg);
}
[data-inverted][data-position~="top"][data-tooltip]:before {
  background: #1D252D;
}
[data-position~="bottom"][data-tooltip]:before {
  -webkit-transform-origin: center bottom;
          transform-origin: center bottom;
}
[data-position~="bottom"][data-tooltip]:after {
  -webkit-transform-origin: center top;
          transform-origin: center top;
}
[data-position="left center"][data-tooltip]:before {
  -webkit-transform-origin: top center;
          transform-origin: top center;
}
[data-position="left center"][data-tooltip]:after {
  -webkit-transform-origin: right center;
          transform-origin: right center;
}
[data-position="right center"][data-tooltip]:before {
  -webkit-transform-origin: right center;
          transform-origin: right center;
}
[data-position="right center"][data-tooltip]:after {
  -webkit-transform-origin: left center;
          transform-origin: left center;
}

/*--------------
    Size
---------------*/

[data-tooltip][data-variation~="mini"]:before,
[data-tooltip][data-variation~="mini"]:after {
  font-size: 0.75rem;
}
[data-tooltip][data-variation~="tiny"]:before,
[data-tooltip][data-variation~="tiny"]:after {
  font-size: 0.83333333rem;
}
[data-tooltip][data-variation~="small"]:before,
[data-tooltip][data-variation~="small"]:after {
  font-size: 0.91666667rem;
}
[data-tooltip][data-variation~="medium"]:before,
[data-tooltip][data-variation~="medium"]:after {
  font-size: 1rem;
}
[data-tooltip][data-variation~="large"]:before,
[data-tooltip][data-variation~="large"]:after {
  font-size: 1.16666667rem;
}
[data-tooltip][data-variation~="huge"]:before,
[data-tooltip][data-variation~="huge"]:after {
  font-size: 1.41666667rem;
}

/*--------------
    Basic
---------------*/

[data-tooltip][data-variation~="basic"]:before {
  display: none;
}

/*--------------
     Spacing
---------------*/

:host(.ui.popup),
.ui.popup {
  margin: 0;
}

/* Extending from Top */
:host(.ui.top.popup),
.ui.top.popup{
  margin: 0 0 0.83333333em;
}
:host(.ui.top.left.popup),
.ui.top.left.popup {
  -webkit-transform-origin: left bottom;
          transform-origin: left bottom;
}
:host(.ui.top.center.popup),
.ui.top.center.popup {
  -webkit-transform-origin: center bottom;
          transform-origin: center bottom;
}
:host(.ui.top.right.popup),
.ui.top.right.popup {
  -webkit-transform-origin: right bottom;
          transform-origin: right bottom;
}

/* Extending from Vertical Center */
:host(.ui.left.center.popup),
.ui.left.center.popup {
  margin: 0 0.83333333em 0 0;
  -webkit-transform-origin: right 50%;
          transform-origin: right 50%;
}
:host(.ui.right.center.popup),
.ui.right.center.popup {
  margin: 0 0 0 0.83333333em;
  -webkit-transform-origin: left 50%;
          transform-origin: left 50%;
}

/* Extending from Bottom */
:host(.ui.bottom.popup),
.ui.bottom.popup {
  margin: 0.83333333em 0 0;
}
:host(.ui.bottom.left.popup),
.ui.bottom.left.popup {
  -webkit-transform-origin: left top;
          transform-origin: left top;
}
:host(.ui.bottom.center.popup),
.ui.bottom.center.popup {
  -webkit-transform-origin: center top;
          transform-origin: center top;
}
:host(.ui.bottom.right.popup),
.ui.bottom.right.popup {
  -webkit-transform-origin: right top;
          transform-origin: right top;
}

/*--------------
     Pointer
---------------*/


/*--- Below ---*/

:host(.ui.bottom.center.popup):before,
.ui.bottom.center.popup:before {
  margin-left: -0.36666667em;
  top: -0.36666667em;
  left: 50%;
  right: auto;
  bottom: auto;
  box-shadow: -1px -1px 0 0 var(--prim-bg);
}
:host(.ui.bottom.left.popup),
.ui.bottom.left.popup {
  margin-left: 0;
}
/*rtl:rename*/
:host(.ui.bottom.left.popup):before,
.ui.bottom.left.popup:before {
  top: -0.36666667em;
  left: 1em;
  right: auto;
  bottom: auto;
  margin-left: 0;
  box-shadow: -1px -1px 0 0 var(--prim-bg);
}
:host(.ui.bottom.right.popup),
.ui.bottom.right.popup {
  margin-right: 0;
}
/*rtl:rename*/
:host(.ui.bottom.right.popup):before,
.ui.bottom.right.popup:before {
  top: -0.36666667em;
  right: 1em;
  bottom: auto;
  left: auto;
  margin-left: 0;
  box-shadow: -1px -1px 0 0 var(--prim-bg);
}

/*--- Above ---*/

:host(.ui.top.center.popup):before,
.ui.top.center.popup:before {
  top: auto;
  right: auto;
  bottom: -0.36666667em;
  left: 50%;
  margin-left: -0.36666667em;
}
:host(.ui.top.left.popup),
.ui.top.left.popup {
  margin-left: 0;
}
/*rtl:rename*/
:host(.ui.top.left.popup):before,
.ui.top.left.popup:before {
  bottom: -0.36666667em;
  left: 1em;
  top: auto;
  right: auto;
  margin-left: 0;
}
:host(.ui.top.right.popup),
.ui.top.right.popup {
  margin-right: 0;
}
/*rtl:rename*/
:host(.ui.top.right.popup):before,
.ui.top.right.popup:before {
  bottom: -0.36666667em;
  right: 1em;
  top: auto;
  left: auto;
  margin-left: 0;
}

/*--- Left Center ---*/

/*rtl:rename*/
:host(.ui.left.center.popup):before,
.ui.left.center.popup:before {
  top: 50%;
  right: -0.36666667em;
  bottom: auto;
  left: auto;
  margin-top: -0.36666667em;
  -webkit-box-shadow: 1px -1px 0 0 var(--prim-bg);
          box-shadow: 1px -1px 0 0 var(--prim-bg);
}

/*--- Right Center  ---*/

/*rtl:rename*/
:host(.ui.right.center.popup):before,
.ui.right.center.popup:before {
  top: 50%;
  left: -0.36666667em;
  bottom: auto;
  right: auto;
  margin-top: -0.36666667em;
  -webkit-box-shadow: -1px 1px 0 0 var(--prim-bg);
          box-shadow: -1px 1px 0 0 var(--prim-bg);
}

/* Arrow Color By Location */
:host(.ui.bottom.popup):before,
.ui.bottom.popup:before {
  background: var(--prim-bg);
}
:host(.ui.right.center.popup):before,
.ui.right.center.popup:before,
:host(.ui.left.center.popup):before,
.ui.left.center.popup:before {
  background: var(--prim-bg);
}
:host(.ui.top.popup):before,
.ui.top.popup:before {
  background: var(--prim-bg);
}

/* Inverted Arrow Color */
:host(.ui.inverted.bottom.popup):before,
.ui.inverted.bottom.popup:before {
  background: #1D252D;
}
:host(.ui.inverted.right.center.popup):before,
.ui.inverted.right.center.popup:before,
:host(.ui.inverted.left.center.popup):before,
.ui.inverted.left.center.popup:before {
  background: #1D252D;
}
:host(.ui.inverted.top.popup):before,
.ui.inverted.top.popup:before {
  background: #1D252D;
}


/*******************************
            Coupling
*******************************/


/* Immediate Nested Grid */
:host(.ui.popup) > .ui.grid:not(.padded),
.ui.popup > .ui.grid:not(.padded) {
  width: calc(100% + 1.75rem);
  margin: -0.7rem -0.875rem;
}


/*******************************
            States
*******************************/

:host(.ui.loading.popup),
.ui.loading.popup {
  display: block;
  visibility: hidden;
  z-index: -1;
}
:host(.ui.animating.popup),
.ui.animating.popup,
:host(.ui.visible.popup),
.ui.visible.popup {
  display: block;
}
:host(.ui.visible.popup),
.ui.visible.popup {
  -webkit-transform: translateZ(0);
          transform: translateZ(0);
  -webkit-backface-visibility: hidden;
          backface-visibility: hidden;
}


/*******************************
            Variations
*******************************/


/*--------------
     Basic
---------------*/

:host(.ui.basic.popup:before),
.ui.basic.popup:before {
  display: none;
}

/*--------------
     Wide
---------------*/

:host(.ui.wide.popup),
.ui.wide.popup {
  max-width: 350px;
}
.ui[class*="very wide"].popup {
  max-width: 550px;
}
@media only screen and (max-width: 767.98px) {
  :host(.ui.wide.popup),
  .ui.wide.popup,
  .ui[class*="very wide"].popup {
    max-width: 250px;
  }
}

/*--------------
     Fluid
---------------*/

:host(.ui.fluid.popup),
.ui.fluid.popup {
  width: 100%;
  max-width: none;
}

/*--------------
     Colors
---------------*/


/* Inverted colors  */
:host(.ui.inverted.popup),
.ui.inverted.popup {
  background: #1D252D;
  color: var(--prim-bg);
  border: none;
  box-shadow: none;
}
:host(.ui.inverted.popup) .header,
.ui.inverted.popup .header {
  background-color: none;
  color: var(--prim-bg);
}
:host(.ui.inverted.popup):before,
.ui.inverted.popup:before {
  background-color: #1D252D;
  box-shadow: none !important;
}

/*--------------
     Flowing
---------------*/

:host(.ui.flowing.popup),
.ui.flowing.popup {
  max-width: none;
}

/*--------------
     Sizes
---------------*/

:host(.ui.mini.popup),
.ui.mini.popup {
  font-size: 0.75rem;
}
:host(.ui.tiny.popup),
.ui.tiny.popup {
  font-size: 0.83333333rem;
}
:host(.ui.small.popup),
.ui.small.popup {
  font-size: 0.91666667rem;
}
:host(.ui.popup),
.ui.popup {
  font-size: 1rem;
}
:host(.ui.large.popup),
.ui.large.popup {
  font-size: 1.16666667rem;
}
:host(.ui.huge.popup),
.ui.huge.popup {
  font-size: 1.41666667rem;
}


/*******************************
         Theme Overrides
*******************************/



/*******************************
        User Overrides
*******************************/

`;
