// @trapezegroup/core-webui-styles/css-variables/sidebar.css.js
export const style = `/*!
 * # Fomantic:host(-UI) - Sidebar
 * http://github.com/fomantic/Fomantic:host(-UI/)
 *
 *
 * Released under the MIT license
 * http://opensource.org/licenses/MIT
 *
 */


/*******************************
            Sidebar
*******************************/


/* Sidebar Menu */
:host(.ui.sidebar) {
  position: fixed;
  top: 0;
  left: 0;
  -webkit-backface-visibility: hidden;
          backface-visibility: hidden;
  -webkit-transition: none;
  transition: none;
  will-change: transform;
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
  visibility: hidden;
  -webkit-overflow-scrolling: touch;
  height: 100% !important;
  max-height: 100%;
  border-radius: 0 !important;
  margin: 0 !important;
  overflow-y: auto !important;
  z-index: 102;
}

/* GPU Layers for Child Elements */
:host(.ui.sidebar) > * {
  -webkit-backface-visibility: hidden;
          backface-visibility: hidden;
}

/*--------------
   Direction
---------------*/

:host(.ui.left.sidebar) {
  right: auto;
  left: 0;
  -webkit-transform: translate3d(-100%, 0, 0);
          transform: translate3d(-100%, 0, 0);
}
:host(.ui.right.sidebar) {
  right: 0 !important;
  left: auto !important;
  -webkit-transform: translate3d(100%, 0, 0);
          transform: translate3d(100%, 0, 0);
}
:host(.ui.top.sidebar),
:host(.ui.bottom.sidebar) {
  width: 100% !important;
  height: auto !important;
}
:host(.ui.top.sidebar) {
  top: 0 !important;
  bottom: auto !important;
  -webkit-transform: translate3d(0, -100%, 0);
          transform: translate3d(0, -100%, 0);
}
:host(.ui.bottom.sidebar) {
  top: auto !important;
  bottom: 0 !important;
  -webkit-transform: translate3d(0, 100%, 0);
          transform: translate3d(0, 100%, 0);
}

/*--------------
     Pushable
---------------*/

.pushable {
  height: 100%;
  overflow-x: hidden;
  padding: 0 !important;
}

/* Whole Page */
body.pushable {
  background: #333F48;
}
body.pushable.dimmed {
  background: inherit;
}

/* Page Context */
.pushable:not(body) {
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
}
.pushable:not(body) > :host(.ui.sidebar),
.pushable:not(body) > .fixed,
.pushable:not(body) > .pusher:after {
  position: absolute;
}

/*--------------
     Fixed
---------------*/

.pushable > .fixed {
  position: fixed;
  -webkit-backface-visibility: hidden;
          backface-visibility: hidden;
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
  will-change: transform;
  z-index: 101;
}

/*--------------
     Page
---------------*/

.pushable > .pusher {
  position: relative;
  -webkit-backface-visibility: hidden;
          backface-visibility: hidden;
  overflow: hidden;
  min-height: 100%;
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
  z-index: 2;
  
/* Pusher should inherit background from context */
  background: inherit;
}
body.pushable > .pusher {
  background: #333F48;
}

/*--------------
     Dimmer
---------------*/

.pushable > .pusher:after {
  position: fixed;
  top: 0;
  right: 0;
  content: '';
  background-color: rgba(0, 0, 0, 0.4);
  overflow: hidden;
  opacity: 0;
  -webkit-transition: opacity 500ms;
  transition: opacity 500ms;
  will-change: opacity;
  z-index: 1000;
}

/*--------------
    Coupling
---------------*/

:host(.ui.sidebar.menu)) .item {
  border-radius: 0 !important;
}


/*******************************
            States
*******************************/


/*--------------
     Dimmed
---------------*/

.pushable > .pusher.dimmed:after {
  width: 100% !important;
  height: 100% !important;
  opacity: 1 !important;
}

/*--------------
    Animating
---------------*/

:host(.ui.animating.sidebar) {
  visibility: visible;
}

/*--------------
     Visible
---------------*/

:host(.ui.visible.sidebar) {
  visibility: visible;
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
}

/* Shadow Direction */
:host(.ui.left.visible.sidebar),
:host(.ui.right.visible.sidebar) {
  -webkit-box-shadow: 0 0 20px #4E5962;
          box-shadow: 0 0 20px #4E5962;
}
:host(.ui.top.visible.sidebar),
:host(.ui.bottom.visible.sidebar) {
  -webkit-box-shadow: 0 0 20px #4E5962;
          box-shadow: 0 0 20px #4E5962;
}

/* Visible On Load */
:host(.ui.visible.left.sidebar) ~ .fixed,
:host(.ui.visible.left.sidebar) ~ .pusher {
  -webkit-transform: translate3d(260px, 0, 0);
          transform: translate3d(260px, 0, 0);
}
:host(.ui.visible.right.sidebar) ~ .fixed,
:host(.ui.visible.right.sidebar) ~ .pusher {
  -webkit-transform: translate3d(-260px, 0, 0);
          transform: translate3d(-260px, 0, 0);
}
:host(.ui.visible.top.sidebar) ~ .fixed,
:host(.ui.visible.top.sidebar) ~ .pusher {
  -webkit-transform: translate3d(0, 36px, 0);
          transform: translate3d(0, 36px, 0);
}
:host(.ui.visible.bottom.sidebar) ~ .fixed,
:host(.ui.visible.bottom.sidebar) ~ .pusher {
  -webkit-transform: translate3d(0, -36px, 0);
          transform: translate3d(0, -36px, 0);
}

/* opposite sides visible forces content overlay */
:host(.ui.visible.left.sidebar) ~ .ui.visible.right.sidebar ~ .fixed,
:host(.ui.visible.left.sidebar) ~ .ui.visible.right.sidebar ~ .pusher,
:host(.ui.visible.right.sidebar) ~ .ui.visible.left.sidebar ~ .fixed,
:host(.ui.visible.right.sidebar) ~ .ui.visible.left.sidebar ~ .pusher {
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
}

/*--------------
       iOS
---------------*/



/*******************************
          Variations
*******************************/


/*--------------
     Width
---------------*/


/* Left / Right */
:host(.ui.thin.left.sidebar),
:host(.ui.thin.right.sidebar) {
  width: 150px;
}
:host(.ui[class*="very) thin"].left.sidebar,
:host(.ui[class*="very) thin"].right.sidebar {
  width: 60px;
}
:host(.ui.left.sidebar),
:host(.ui.right.sidebar) {
  width: 260px;
}
:host(.ui.wide.left.sidebar),
:host(.ui.wide.right.sidebar) {
  width: 350px;
}
:host(.ui[class*="very) wide"].left.sidebar,
:host(.ui[class*="very) wide"].right.sidebar {
  width: 475px;
}

/* Left Visible */
:host(.ui.visible.thin.left.sidebar) ~ .fixed,
:host(.ui.visible.thin.left.sidebar) ~ .pusher {
  -webkit-transform: translate3d(150px, 0, 0);
          transform: translate3d(150px, 0, 0);
}
:host(.ui.visible[class*="very) thin"].left.sidebar ~ .fixed,
:host(.ui.visible[class*="very) thin"].left.sidebar ~ .pusher {
  -webkit-transform: translate3d(60px, 0, 0);
          transform: translate3d(60px, 0, 0);
}
:host(.ui.visible.wide.left.sidebar) ~ .fixed,
:host(.ui.visible.wide.left.sidebar) ~ .pusher {
  -webkit-transform: translate3d(350px, 0, 0);
          transform: translate3d(350px, 0, 0);
}
:host(.ui.visible[class*="very) wide"].left.sidebar ~ .fixed,
:host(.ui.visible[class*="very) wide"].left.sidebar ~ .pusher {
  -webkit-transform: translate3d(475px, 0, 0);
          transform: translate3d(475px, 0, 0);
}

/* Right Visible */
:host(.ui.visible.thin.right.sidebar) ~ .fixed,
:host(.ui.visible.thin.right.sidebar) ~ .pusher {
  -webkit-transform: translate3d(-150px, 0, 0);
          transform: translate3d(-150px, 0, 0);
}
:host(.ui.visible[class*="very) thin"].right.sidebar ~ .fixed,
:host(.ui.visible[class*="very) thin"].right.sidebar ~ .pusher {
  -webkit-transform: translate3d(-60px, 0, 0);
          transform: translate3d(-60px, 0, 0);
}
:host(.ui.visible.wide.right.sidebar) ~ .fixed,
:host(.ui.visible.wide.right.sidebar) ~ .pusher {
  -webkit-transform: translate3d(-350px, 0, 0);
          transform: translate3d(-350px, 0, 0);
}
:host(.ui.visible[class*="very) wide"].right.sidebar ~ .fixed,
:host(.ui.visible[class*="very) wide"].right.sidebar ~ .pusher {
  -webkit-transform: translate3d(-475px, 0, 0);
          transform: translate3d(-475px, 0, 0);
}


/*******************************
          Animations
*******************************/


/*--------------
    Overlay
---------------*/


/* Set-up */
:host(.ui.overlay.sidebar) {
  z-index: 102;
}

/* Initial */
:host(.ui.left.overlay.sidebar) {
  -webkit-transform: translate3d(-100%, 0, 0);
          transform: translate3d(-100%, 0, 0);
}
:host(.ui.right.overlay.sidebar) {
  -webkit-transform: translate3d(100%, 0, 0);
          transform: translate3d(100%, 0, 0);
}
:host(.ui.top.overlay.sidebar) {
  -webkit-transform: translate3d(0%, -100%, 0);
          transform: translate3d(0%, -100%, 0);
}
:host(.ui.bottom.overlay.sidebar) {
  -webkit-transform: translate3d(0%, 100%, 0);
          transform: translate3d(0%, 100%, 0);
}

/* Animation */
.animating:host(.ui.overlay.sidebar),
:host(.ui.visible.overlay.sidebar) {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
}

/* End - Sidebar */
:host(.ui.visible.left.overlay.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}
:host(.ui.visible.right.overlay.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}
:host(.ui.visible.top.overlay.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}
:host(.ui.visible.bottom.overlay.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}

/* End - Pusher */
:host(.ui.visible.overlay.sidebar) ~ .fixed,
:host(.ui.visible.overlay.sidebar) ~ .pusher {
  -webkit-transform: none !important;
          transform: none !important;
}

/*--------------
      Push
---------------*/


/* Initial */
:host(.ui.push.sidebar) {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
  z-index: 102;
}

/* Sidebar - Initial */
:host(.ui.left.push.sidebar) {
  -webkit-transform: translate3d(-100%, 0, 0);
          transform: translate3d(-100%, 0, 0);
}
:host(.ui.right.push.sidebar) {
  -webkit-transform: translate3d(100%, 0, 0);
          transform: translate3d(100%, 0, 0);
}
:host(.ui.top.push.sidebar) {
  -webkit-transform: translate3d(0%, -100%, 0);
          transform: translate3d(0%, -100%, 0);
}
:host(.ui.bottom.push.sidebar) {
  -webkit-transform: translate3d(0%, 100%, 0);
          transform: translate3d(0%, 100%, 0);
}

/* End */
:host(.ui.visible.push.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}

/*--------------
    Uncover
---------------*/


/* Initial */
:host(.ui.uncover.sidebar) {
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
  z-index: 1;
}

/* End */
:host(.ui.visible.uncover.sidebar) {
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
}

/*--------------
   Slide Along
---------------*/


/* Initial */
:host(.ui.slide.along.sidebar) {
  z-index: 1;
}

/* Sidebar - Initial */
:host(.ui.left.slide.along.sidebar) {
  -webkit-transform: translate3d(-50%, 0, 0);
          transform: translate3d(-50%, 0, 0);
}
:host(.ui.right.slide.along.sidebar) {
  -webkit-transform: translate3d(50%, 0, 0);
          transform: translate3d(50%, 0, 0);
}
:host(.ui.top.slide.along.sidebar) {
  -webkit-transform: translate3d(0, -50%, 0);
          transform: translate3d(0, -50%, 0);
}
:host(.ui.bottom.slide.along.sidebar) {
  -webkit-transform: translate3d(0%, 50%, 0);
          transform: translate3d(0%, 50%, 0);
}

/* Animation */
:host(.ui.animating.slide.along.sidebar) {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
}

/* End */
:host(.ui.visible.slide.along.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}

/*--------------
   Slide Out
---------------*/


/* Initial */
:host(.ui.slide.out.sidebar) {
  z-index: 1;
}

/* Sidebar - Initial */
:host(.ui.left.slide.out.sidebar) {
  -webkit-transform: translate3d(50%, 0, 0);
          transform: translate3d(50%, 0, 0);
}
:host(.ui.right.slide.out.sidebar) {
  -webkit-transform: translate3d(-50%, 0, 0);
          transform: translate3d(-50%, 0, 0);
}
:host(.ui.top.slide.out.sidebar) {
  -webkit-transform: translate3d(0%, 50%, 0);
          transform: translate3d(0%, 50%, 0);
}
:host(.ui.bottom.slide.out.sidebar) {
  -webkit-transform: translate3d(0%, -50%, 0);
          transform: translate3d(0%, -50%, 0);
}

/* Animation */
:host(.ui.animating.slide.out.sidebar) {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
}

/* End */
:host(.ui.visible.slide.out.sidebar) {
  -webkit-transform: translate3d(0%, 0, 0);
          transform: translate3d(0%, 0, 0);
}

/*--------------
   Scale Down
---------------*/


/* Initial */
:host(.ui.scale.down.sidebar) {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
  z-index: 102;
}

/* Sidebar - Initial  */
:host(.ui.left.scale.down.sidebar) {
  -webkit-transform: translate3d(-100%, 0, 0);
          transform: translate3d(-100%, 0, 0);
}
:host(.ui.right.scale.down.sidebar) {
  -webkit-transform: translate3d(100%, 0, 0);
          transform: translate3d(100%, 0, 0);
}
:host(.ui.top.scale.down.sidebar) {
  -webkit-transform: translate3d(0%, -100%, 0);
          transform: translate3d(0%, -100%, 0);
}
:host(.ui.bottom.scale.down.sidebar) {
  -webkit-transform: translate3d(0%, 100%, 0);
          transform: translate3d(0%, 100%, 0);
}

/* Pusher - Initial */
:host(.ui.scale.down.left.sidebar) ~ .pusher {
  -webkit-transform-origin: 75% 50%;
          transform-origin: 75% 50%;
}
:host(.ui.scale.down.right.sidebar) ~ .pusher {
  -webkit-transform-origin: 25% 50%;
          transform-origin: 25% 50%;
}
:host(.ui.scale.down.top.sidebar) ~ .pusher {
  -webkit-transform-origin: 50% 75%;
          transform-origin: 50% 75%;
}
:host(.ui.scale.down.bottom.sidebar) ~ .pusher {
  -webkit-transform-origin: 50% 25%;
          transform-origin: 50% 25%;
}

/* Animation */
:host(.ui.animating.scale.down)) > .visible.ui.sidebar {
  -webkit-transition: -webkit-transform 500ms ease;
  transition: -webkit-transform 500ms ease;
  transition: transform 500ms ease;
  transition: transform 500ms ease, -webkit-transform 500ms ease;
}
:host(.ui.visible.scale.down.sidebar) ~ .pusher,
:host(.ui.animating.scale.down.sidebar) ~ .pusher {
  display: block !important;
  width: 100%;
  height: 100%;
  overflow: hidden !important;
}

/* End */
:host(.ui.visible.scale.down.sidebar) {
  -webkit-transform: translate3d(0, 0, 0);
          transform: translate3d(0, 0, 0);
}
:host(.ui.visible.scale.down.sidebar) ~ .pusher {
  -webkit-transform: scale(0.75);
          transform: scale(0.75);
}


/*******************************
         Theme Overrides
*******************************/



/*******************************
         Site Overrides
*******************************/

`;
