// @trapezegroup/core-wc-layout/src/components/tabs/core-tab-content.js
/**  import litElement base class and html helper function */
import { LitElement, html, css } from 'lit-element';

/** See core-layout.js for examples. */
export class CoreTabContent extends LitElement {
    /** @inheritdoc */
    static get styles() {
        return [
            css`
                :host {
                    display: none;
                    background-color: var(--tab-content-bg, var(--prim-bg));
                }

                :host([active]) {
                    display: block;
                    height: 100%;
                }
            `,
        ];
    }

    /** @inheritdoc initialize as content */
    connectedCallback() {
        super.connectedCallback();

        this.setAttribute('slot', 'content');
    }

    /** We can initialize the properties here in the constructor */
    constructor() {
        // Must call superconstructor first.
        super();
    }

    /** @inheritdoc */
    render() {
        return html`<slot></slot>`;
    }
}
customElements.define('core-tab-content', CoreTabContent);
