// @trapezegroup/core-webui-styles/dist/theme-default/cssJavaScript/site.css.js
export const style = `/*!
 * # Fomantic-UI - Site
 * http://github.com/fomantic/Fomantic-UI/
 *
 *
 * Released under the MIT license
 * http://opensource.org/licenses/MIT
 *
 */


/*******************************
             Page
*******************************/

@import url('https://fonts.googleapis.com/css?family=Source Sans Pro:400,700,400italic,700italic&subset=latin&display=swap');
html,
body {
  height: 100%;
}
html {
  font-size: 12px;
}
body {
  margin: 0;
  padding: 0;
  overflow-x: hidden;
  min-width: 320px;
  background: #FFFFFF;
  font-family: 'Source Sans Pro', 'Segoe UI', 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-size: 12px;
  line-height: 1.4285em;
  color: rgba(0, 0, 0, 0.87);
}


/*******************************
             Headers
*******************************/

h1,
h2,
h3,
h4,
h5 {
  font-family: 'Source Sans Pro', 'Segoe UI', 'Helvetica Neue', Arial, Helvetica, sans-serif;
  line-height: 1.28571429em;
  margin: calc(2rem - 0.1428571428571429em) 0 1rem;
  font-weight: bold;
  padding: 0;
}
h1 {
  min-height: 1rem;
  font-size: 2rem;
}
h2 {
  font-size: 1.71428571rem;
}
h3 {
  font-size: 1.28571429rem;
}
h4 {
  font-size: 1.07142857rem;
}
h5 {
  font-size: 1rem;
}
h1:first-child,
h2:first-child,
h3:first-child,
h4:first-child,
h5:first-child {
  margin-top: 0;
}
h1:last-child,
h2:last-child,
h3:last-child,
h4:last-child,
h5:last-child {
  margin-bottom: 0;
}


/*******************************
             Text
*******************************/

p {
  margin: 0 0 1em;
  line-height: 1.4285em;
}
p:first-child {
  margin-top: 0;
}
p:last-child {
  margin-bottom: 0;
}

/*-------------------
        Links
--------------------*/

a {
  color: #4183C4;
  text-decoration: none;
}
a:hover {
  color: #1e70bf;
  text-decoration: none;
}


/*******************************
         Scrollbars
*******************************/



/*******************************
          Highlighting
*******************************/


/* Site */
::-webkit-selection {
  background-color: #CCE2FF;
  color: rgba(0, 0, 0, 0.87);
}
::-moz-selection {
  background-color: #CCE2FF;
  color: rgba(0, 0, 0, 0.87);
}
::selection {
  background-color: #CCE2FF;
  color: rgba(0, 0, 0, 0.87);
}

/* Form */
textarea::-webkit-selection,
input::-webkit-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: rgba(0, 0, 0, 0.87);
}
textarea::-moz-selection,
input::-moz-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: rgba(0, 0, 0, 0.87);
}
textarea::-moz-selection, input::-moz-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: rgba(0, 0, 0, 0.87);
}
textarea::selection,
input::selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: rgba(0, 0, 0, 0.87);
}

/* Force Simple Scrollbars */
body ::-webkit-scrollbar {
  -webkit-appearance: none;
  width: 10px;
  height: 10px;
}
body ::-webkit-scrollbar-track {
  background: rgba(0, 0, 0, 0.1);
  border-radius: 0;
}
body ::-webkit-scrollbar-thumb {
  cursor: pointer;
  border-radius: 5px;
  background: rgba(0, 0, 0, 0.25);
  -webkit-transition: color 0.2s ease;
  transition: color 0.2s ease;
}
body ::-webkit-scrollbar-thumb:window-inactive {
  background: rgba(0, 0, 0, 0.15);
}
body ::-webkit-scrollbar-thumb:hover {
  background: rgba(128, 135, 139, 0.8);
}

/* Inverted UI */
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-track {
  background: rgba(255, 255, 255, 0.1);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb {
  background: rgba(255, 255, 255, 0.25);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb:window-inactive {
  background: rgba(255, 255, 255, 0.15);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb:hover {
  background: rgba(255, 255, 255, 0.35);
}


/*******************************
        Global Overrides
*******************************/



/*******************************
           Overrides
*******************************/

:root {
  
/* Make font size 10px so that rem can be used in app */
  --core-size-base: 62.5%;
  font-size: var(--core-size-base);
  --primary-background-color: #FFFFFF;
  --primary-font-color: rgba(0, 0, 0, 0.87);
  --secondary-background-color: #E4EFF2;
  --secondary-font-color: #5F757F;
  --primary-border-color: rgba(34, 36, 38, 0.15);
  --primary-on-color: #1d999f;
  --primary-off-color: #8c9ea7;
  
/* Status Colours */
  --primary-status-good-font: #396A07;
  --primary-status-good-bg: #FDFFFA;
  --primary-status-good-border: #61A220;
  --primary-status-bad-font: #950516;
  --primary-status-bad-bg: #F8F2F3;
  --primary-status-bad-border: #B30015;
  --primary-status-attention-font: #F0B723;
  --primary-status-attention-bg: transparent;
  
/* not defined yet */
  --primary-status-attention-border: #F0B723;
  --primary-status-unknown-font: #087D88;
  --primary-status-unknown-bg: #F2F8F8;
  --primary-status-unknown-border: #087D88;
  

/*******************************
        DEFAULT THEME
*******************************/

  
/* BRAND COLOURS */
  --col-1: #333F48;
  
/* brand gray blue */
  --col-1-hs: 206,17%;
  
/* brand gray blue - hs values*/
  --col-2: #07828D;
  /* teal */
  --col-2-hs: 185,91%;
  
/* teal h s values - hs values*/
  --col-3: #007099;
  /* accessible blue */
  --col-3-hs: 196,100%;
  
/* accessible blue - hs values*/
  --col-4: #C70505;
  /* error red, critical icon */
  --col-4-hs: 0, 95%;
  
/* error red, critical icon - hs values*/
  --col-5: #8C9EA7;
  /* gray */
  --col-6: #5B727C;
  
/* bg and button text */
  --col-7: #D1E9F2;
  
/* selected row */
  --col-8: #EEF2F4;
  
/* bg */
  --col-9: #8A1538;
  
/* old brand burgundy */
  --col-10: #00A3E0;
  
/* old brand blue */
  --col-10-hs: 196, 100%;
  
/* old brand blue - hs values*/
  --col-11: #FF9E18;
  /* old brand orange */
  --col-11-hs: 35,100%;
  
/* old brand orange - hs values*/
  --col-12: #338533;
  /* new button green */
  --col-12-hs: 120, 45%;
  
/* new button green - hs values*/
  --col-13: #A4BCC6;
  /* inactive icons */
  --col-14: #1D2429;
  
/* darkest bg */
  --col-15: #FFFFFF;
  
/* white */
  --col-16: #DCE5E9;
  
/* nav text */
  --col-17: #D7D9DB;
  
/* hover for teal */
  --col-18: #C8D9E3;
  
/* YM divider */
  --col-18-hs: 202,33%;
  
/* YM divider */
  --col-20: #F03030;
  
/* new light red */
  --col-20-hs: 0, 86%;
  
/* new light red - hs values*/
  --col-21: #2FC4BE;
  /* new light teal */
  --col-21-hs: 178, 61%;
  
/* new light teal - hs values*/
  --col-22: #A4A0EE;
  /* new light purple */
  --col-22-hs: 243,70%;
  
/* new light purple - hs values*/
  --col-23: #825AE2;
  /* new button purple */
  --col-23-hs: 258, 70%;
  
/* new button purple - hs values*/
  --col-24: #C25700;
  /* new button orange */
  --col-24-hs: 27, 100%;
  
/* new button orange - hs values*/
  --col-25: #87BE50;
  /* new light green */
  --col-25-hs: 90, 46%;
  
/* new light green - hs values*/
  

/*******************************/

  /* BACKGROUNDS */
  --prim-bg: #EEF2F4;
  --sec-bg: #FFFFFF;
  --sec-bg-rgb: 29,36,41;
  
/* 100%; ffffff */
  --ter-bg: #1D2429;
  
/* darkest bg */
  --quat-bg: #FDFDFD;
  /*YM*/
  --quin-bg: #A4BCC6;
  --quin-bg-rgb: 164,188,198;
  
/* 20%; #A4BCC6 */
  --sen-bg: #5B727C;
  --sep-bg: #DCE5E9;
  --sel-bg: hsl(var(--col-21-hs), 60%);
  --nav-bg: #333F48;
  --nav-sel-bg: hsl(var(--col-3-hs), 23%);
  --prim-sel-bg: hsl(var(--col-21-hs), 60%);
  
/* selected row */
  --prim-dropdown-bg: #ffffff;
  --prim-confirm-bg: #07838E;
  --prim-error-bg: #B30015;
  --sec-cal-pic: #FFFFFF;
  --pop-up: #FFFFFF;
  --off-bg: #8C9EA7;
  
/* ie gray toggle */
  --prim-shdw: rgba(51, 63, 72, 0.4) 0px 3px 6px 0px;
  
/* #333F48 */
  --prim-bg-hover-rgb: 51,63,72;
  
/* 10%; calendar #333F48 */
  

/*******************************/

  
/* DIVIDER */
  --prim-divider: #DCE5E9;
  --prim-divider-rgba: 220,229,233;
  
/* 40% #A4BCC6; */
  --prim-divider-grid: #FFFFFF;
  --sec-divider: #ffffff;
  --ter-divider: #EEF2F4;
  --quat-divider: #C8D9E3;
  /*YM APA*/
  --quat-divider-rgb: hsl(var(--col-18-hs), 50%);
  --prim-input-divider: #DCE5E9;
  
/* OUTLINES */
  --prim-brd: #DCE5E9;
  --prim-sel-brd: #A4BCC6;
  

/*******************************/

  
/* TEXT */
  --prim-txt: #333F48;
  --prim-txt-hover: hsl(var(--col-1-hs), 26%);
  --prim-txt-error: #C70505;
  --sec-txt: #5B727C;
  --nav-txt: #FFFFFF;
  
/* TEXT LINK */
  --prim-link-txt: #5B727C;
  --prim-link-txt-hover: #333F48;
  

/******************************* /
 
 
/* BUTTONS */

  --prim-button-bg: #07828D;
  
/* teal */
  --sec-button-bg: transparent;
  --ter-button-bg: #333F48;
  
/* dark */
  --prim-button-border: transparent;
  --sec-button-border: #A4BCC6;
  
/* grey */
  --ter-button-border: #333F48;
  
/* dark */
  --prim-button-txt: #ffffff;
  --sec-button-txt: #5B727C;
  
/* D grey */
  --ter-button-txt: #333F48;
  
/* dark */
  
/* BUTTON COLOR - HOVER */
  --prim-button-bg-hover: hsl(var(--col-2-hs), 30%);
  --sec-button-bg-hover: #333F48;
  
/* dark */
  --sec-button-bg-hover: hsl(var(--col-1-hs), 26%);
  --sec-button-bg-hover-border: hsl(var(--col-1-hs), 26%);
  

/*******************************/

  
/* ICONS */
  --prim-icon: #5B727C;
  --prim-icon-bg: #FFFFFF;
  --prim-icon-highlight: #07828D;
  --sec-icon-bg: #A4BCC6;
  --prim-icon-off: #5B727C;
  --prim-icon-hover: #333F48;
  --sec-icon-off: #338533;
  --sec-icon-hover: hsl(var(--col-12-hs), 70%);
  --ter-icon-off: #D7D9DB;
  --ter-icon-hover: #FF9E18;
  --qua-icon-light: #DCE5E9;
  

/*******************************/

  
/* SEQUENCE HIERARCHY
1 primary   --  prim
2 secondary --  sec
3 tertiary  --  ter
4 quaternary--  quat
5 quinary   --  quin
6 senary    --  sen
7 septenary --  sep
8 octonary  --  oct
9 nonary    --  non
10 denary   --  den
*/
  

/*******************************/

  /* YM */
  

/*******************************/

  /*BACKGROUNDS*/
  --tri-bg: hsl(var(--col-2), 20%);
  
/* selected nav */
  -bg--prim-bg-status: #338533;
  
/* green */
  --prim-border-status: #87BE50;
  
/* green */
  --prim-txt-status: #87BE50;
  
/* green */
  --prim-icon-status: #338533;
  
/* green */
  --prim-confirm-bg: #07828D;
  --prim-error: #C70505;
  --prim-error-bg: #C70505;
  --prim-dialog-bg: #EEF2F4;
  --sec-bg-status: #C70505;
  
/* red */
  --sec-border-status: #C70505;
  
/* red */
  --sec-txt-status: #5B727C;
  
/* red */
  --sec-icon-status: #C70505;
  
/* red */
  --ter-bg-status: #007099;
  
/* blue */
  --ter-border-status: #007099;
  
/* blue */
  --ter-txt-status: #007099;
  
/* blue */
  --ter-icon-status: #007099;
  
/* blue */
  
/* ICONS */
  --prim-icon-YM: #099FAC;
  --sec-icon-YM: #F0B723;
  

/*******************************/

}


/*******************************
         Site Overrides
*******************************/

`;
