// @trapezegroup/core-webui-styles/dist/theme-dark/cssJavaScript/site.css.js
export const style = `/*!
 * # Fomantic-UI - Site
 * http://github.com/fomantic/Fomantic-UI/
 *
 *
 * Released under the MIT license
 * http://opensource.org/licenses/MIT
 *
 */


/*******************************
             Page
*******************************/

@import url('https://fonts.googleapis.com/css?family=Source Sans Pro:400,700,400italic,700italic&subset=latin&display=swap');
html,
body {
  height: 100%;
}
html {
  font-size: 12px;
}
body {
  margin: 0;
  padding: 0;
  overflow-x: hidden;
  min-width: 320px;
  background: #333F48;
  font-family: 'Source Sans Pro', 'Segoe UI', 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-size: 12px;
  line-height: 1.4285em;
  color: #B0C7D1;
}


/*******************************
             Headers
*******************************/

h1,
h2,
h3,
h4,
h5 {
  font-family: 'Source Sans Pro', 'Segoe UI', 'Helvetica Neue', Arial, Helvetica, sans-serif;
  line-height: 1.28571429em;
  margin: calc(2rem - 0.1428571428571429em) 0 1rem;
  font-weight: bold;
  padding: 0;
}
h1 {
  min-height: 1rem;
  font-size: 2rem;
}
h2 {
  font-size: 1.71428571rem;
}
h3 {
  font-size: 1.28571429rem;
}
h4 {
  font-size: 1.07142857rem;
}
h5 {
  font-size: 1rem;
}
h1:first-child,
h2:first-child,
h3:first-child,
h4:first-child,
h5:first-child {
  margin-top: 0;
}
h1:last-child,
h2:last-child,
h3:last-child,
h4:last-child,
h5:last-child {
  margin-bottom: 0;
}


/*******************************
             Text
*******************************/

p {
  margin: 0 0 1em;
  line-height: 1.4285em;
}
p:first-child {
  margin-top: 0;
}
p:last-child {
  margin-bottom: 0;
}

/*-------------------
        Links
--------------------*/

a {
  color: #4183C4;
  text-decoration: none;
}
a:hover {
  color: #1e70bf;
  text-decoration: none;
}


/*******************************
         Scrollbars
*******************************/



/*******************************
          Highlighting
*******************************/


/* Site */
::-webkit-selection {
  background-color: #CCE2FF;
  color: #B0C7D1;
}
::-moz-selection {
  background-color: #CCE2FF;
  color: #B0C7D1;
}
::selection {
  background-color: #CCE2FF;
  color: #B0C7D1;
}

/* Form */
textarea::-webkit-selection,
input::-webkit-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: #B0C7D1;
}
textarea::-moz-selection,
input::-moz-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: #B0C7D1;
}
textarea::-moz-selection, input::-moz-selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: #B0C7D1;
}
textarea::selection,
input::selection {
  background-color: rgba(100, 100, 100, 0.4);
  color: #B0C7D1;
}

/* Force Simple Scrollbars */
body ::-webkit-scrollbar {
  -webkit-appearance: none;
  width: 10px;
  height: 10px;
}
body ::-webkit-scrollbar-track {
  background: rgba(0, 0, 0, 0.1);
  border-radius: 0;
}
body ::-webkit-scrollbar-thumb {
  cursor: pointer;
  border-radius: 5px;
  background: rgba(0, 0, 0, 0.25);
  -webkit-transition: color 0.2s ease;
  transition: color 0.2s ease;
}
body ::-webkit-scrollbar-thumb:window-inactive {
  background: rgba(0, 0, 0, 0.15);
}
body ::-webkit-scrollbar-thumb:hover {
  background: rgba(128, 135, 139, 0.8);
}

/* Inverted UI */
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-track {
  background: rgba(255, 255, 255, 0.1);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb {
  background: rgba(255, 255, 255, 0.25);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb:window-inactive {
  background: rgba(255, 255, 255, 0.15);
}
body .ui.inverted:not(.dimmer)::-webkit-scrollbar-thumb:hover {
  background: rgba(255, 255, 255, 0.35);
}


/*******************************
        Global Overrides
*******************************/



/*******************************
           Overrides
*******************************/

:root {
  --core-size-base: 62.5%;
  font-size: var(--core-size-base);
  --primary-background-color: #333F48;
  --primary-font-color: #B0C7D1;
  --secondary-background-color: #2B353C;
  --secondary-font-color: #FFF;
  --primary-border-color: #4E5962;
  --primary-on-color: #099FAC;
  --primary-off-color: var(--primary-font-color);
  
/* Status Colours */
  --primary-status-good-font: #27E28B;
  --primary-status-good-bg: transparent;
  --primary-status-good-border: #27E28B;
  --primary-status-bad-font: #FF7B8A;
  --primary-status-bad-bg: transparent;
  --primary-status-bad-border: #F86173;
  --primary-status-attention-font: #F0B723;
  --primary-status-attention-bg: transparent;
  --primary-status-attention-border: #F0B723;
  --primary-status-unknown-font: #00DCEF;
  --primary-status-unknown-bg: transparent;
  --primary-status-unknown-border: #00DCEF;
  

/*******************************
          DARK THEME
*******************************/

  --col-1: #333F48;
  
/* brand gray blue */
  --col-2: #099FAC;
  
/* teal */
  --col-3: #007099;
  
/* accessible blue */
  --col-4: #F03030;
  
/* error red, critical icon */
  --col-5: #000000;
  
/* icons */
  --col-6: #A4BCC6;
  
/* button text/inactive icons*/
  --col-7: #D1E9F2;
  /* selected row */
  --col-8: #2C373E;
  
/* bg */
  --col-9: #87BE50;
  
/* brand burgundy */
  --col-10: #00A3E0;
  
/* brand blue */
  --col-11: #ED8B00;
  
/* brand orange */
  --col-12: #B6BF00;
  
/* brand green */
  --col-13: #5B727C;
  
/* btn border */
  --col-14: #1D2429;
  
/* sec bg, inactive icons, selected row, */
  --col-15: #ffffff;
  
/* white, active icons */
  --col-16: #DCE5E9;
  
/* nav text, divider */
  --col-18: #C8D9E3;
  
/* YM divider */
  --col-19: #222B30;
  
/* YM ribbon, empty asset cell */
  --col-20: #F03030;
  
/* new light red */
  --col-21: #2FC4BE;
  
/* new light teal */
  --col-22: #A4A0EE;
  
/* new light purple */
  --col-23: #825AE2;
  
/* new button purple */
  --col-24: #FF9E18;
  
/* new button orange */
  --col-25: #87BE50;
  
/* new light green */
  

/*******************************/

  
/* BACKGROUNDS */
  --prim-bg: #2C373E;
  --sec-bg: #1D2429;
  --sec-bg: hsla(var(--col-14), 17%);
  
/* tripper bg, YM ribbion grid  */
  --sen-bg: #5B727C;
  --sel-bg: #2C373E;
  --nav-bg: #2C373E;
  --nav-sel-bg: #1D2429;
  --prim-sel-bg: #1D2429;
  
/* selected row */
  --prim-dropdown-bg: #2C373E;
  --prim-dialog-bg: #2C373E;
  --prim-confirm-bg: #07838E;
  --prim-error-bg: #F03030;
  --sec-cal-pic: #2C373E;
  --pop-up: #2C373E;
  --off-bg: #A4BCC6;
  
/* ie gray toggle */
  --prim-shdw: #1D2429;
  --prim-bg-hover: hsla(var(--col-14), 0.4);
  /*calendar rollover */
  

/*******************************/

  
/* DIVIDER */
  --prim-divider: hsla(var(--col-8), 23%);
  --sec-divider: #1D2429;
  
/* OUTLINES */
  --prim-brd: hsla(var(--col-8), 35%);
  --prim-sel-brd: #099FAC;
  

/*******************************/

  
/* TEXT */
  --prim-txt: #ffffff;
  --prim-txt-hover: hsl(var(--col-15), 50%);
  --prim-txt-error: #F03030;
  --sec-txt: #A4BCC6;
  
/* TEXT LINK */
  --prim-link-txt: #A4BCC6;
  --prim-link-txt-hover: #ffffff;
  

/*******************************   /
 
 
/* BUTTONS */

  --prim-button-bg: #099FAC;
  
/* teal */
  --sec-button-bg: transparent;
  --ter-button-bg: #ffffff;
  
/* white */
  --prim-button-border: transparent;
  --sec-button-border: #5B727C;
  
/* D grey */
  --ter-button-border: #ffffff;
  
/* white */
  --prim-button-txt: #ffffff;
  --prim-button-txt-rgb: 255,255,255;
  
/* 0%; #ffffff */
  --sec-button-txt: #A4BCC6;
  
/* grey */
  --ter-button-txt: #ffffff;
  
/* white */
  
/* BUTTON COLOR - HOVER */
  --prim-button-bg-hover: hsl(var(--col-2), 26%);
  --sec-button-bg-hover: transparent;
  --sec-button-bg-hover-border: #A4BCC6;
  
/* grey */
  

/*******************************  /
 

/* ICONS */

  --prim-icon-bg: #A4BCC6;
  --sec-icon-bg: #FFFFFF;
  --ter-icon-bg: #000000;
  --prim-icon-off: #1D2429;
  --prim-icon-hover: #A4BCC6;
  --sec-icon-off: #1D2429;
  --sec-icon-hover: #87BE50;
  

/*******************************/

  
/* YM */
  

/*******************************/

  /*BACKGROUNDS*/
  --prim-bg-status: #2C373E;
  
/* green */
  --prim-border-status: #87BE50;
  
/* green */
  --prim-txt-status: #87BE50;
  
/* green */
  --prim-icon-status: #87BE50;
  
/* green */
  --sec-bg-status: #2C373E;
  
/* red */
  --sec-border-status: #F03030;
  
/* red */
  --sec-txt-status: #F03030;
  
/* red */
  --sec-icon-status: #F03030;
  
/* red */
  --ter-bg-status: #2C373E;
  
/* blue */
  --ter-border-status: #2FC4BE;
  
/* blue */
  --ter-txt-status: #2FC4BE;
  
/* blue */
  --ter-icon-status: #2FC4BE;
  
/* blue */
  
/* ICONS */
  --prim-icon-YM: #099FAC;
  --sec-icon-YM: #FF9E18;
  

/*******************************/

}


/*******************************
         Site Overrides
*******************************/

`;
