// @trapezegroup/core-wc-inputs/src/components/deprecated/core-input-boolean.js
/* eslint-disable require-jsdoc */
// @ts-check
import { TcfBoolean } from '/node_modules/@trapezegroup/core-js-tcftypes/src/lib/types.js';
import { CoreInputBase } from './core-input-base.js';

export class CoreInputBoolean extends CoreInputBase {
    static CONTAINER_TAG = 'INPUT';
    static DEFAULT_DATATYPE = TcfBoolean;

    getDataType() {
        this.dataType.setValue(this.isChecked());
        return this.dataType;
    }

    show() {
        super.show();
        this.focus();
        // TODO We overide select in common.js to select element not user selection. Review this code.
        // @ts-ignore
        this.select();
    }

    isChecked() {
        return this.checked;
    }
    getSelectedValue() {
        // @ts-ignore
        return `${this.getInputElement().checked}`;
    }

    getState() {
        return { value: this.checked };
    }

    setState(state) {
        if (state !== undefined) this.checked = state.value;
    }
}

customElements.define('core-input-boolean', CoreInputBoolean);
