// @trapezegroup/core-wc-workspace/src/components/workspace/core-workspace-footer.js
/**  import litElement base class and html helper function */
import { LitElement, html, css } from 'lit-element';

/** This is an example of a component which extends the litElement */
export class CoreWorkspaceFooter extends LitElement {
    /** @inheritdoc */
    static get styles() {
        return [
            css`
                :host {
                    display: block;
                }
                .footer {
                    position: absolute;
                    bottom: 0;
                    width: 100%;
                    background-color: var(--prim-txt);
                    height: 4.2rem;
                    color: var(--prim-button-txt);
                    padding: 0.8rem;
                    box-sizing: border-box;
                }
            `,
        ];
    }

    /** We can initialize the properties here in the constructor */
    constructor() {
        // Must call superconstructor first.
        super();
    }

    /** @inheritdoc */
    render() {
        return html`
        <slot></slot>
        <div class="footer">
            <slot name="footer"></slot>
        </div>
    `;
    }
}
customElements.define('core-workspace-footer', CoreWorkspaceFooter);
