// @trapezegroup/core-webui-styles/css-variables/table.css.js
export const styles = `/*!
 * # Fomantic-UI - Table
 * http://github.com/fomantic/Fomantic-UI/
 *
 *
 * Released under the MIT license
 * http://opensource.org/licenses/MIT
 *
 */


/*******************************
             Table
*******************************/


/* Prototype */
.ui.table {
  width: 100%;
  color: var(--prim-bg);
  margin: 1em 0;
  border: 1px solid var(--prim-brd);
  -webkit-box-shadow: none;
          box-shadow: none;
  border-radius: 0.33333333rem;
  text-align: left;
  vertical-align: middle;
  color: var(--prim-txt);
  border-collapse: separate;
  border-spacing: 0;
}
.ui.table:first-child {
  margin-top: 0;
}
.ui.table:last-child {
  margin-bottom: 0;
}
.ui.table > thead,
.ui.table > tbody {
  text-align: inherit;
  vertical-align: inherit;
}


/*******************************
             Parts
*******************************/


/* Table Content */
.ui.table th,
.ui.table td {
  -webkit-transition: background 0.1s ease, color 0.1s ease;
  transition: background 0.1s ease, color 0.1s ease;
}

/* Headers */
.ui.table > thead {
  -webkit-box-shadow: none;
          box-shadow: none;
}
.ui.table > thead > tr > th {
  cursor: auto;
  background: var(--sec-bg);
  text-align: inherit;
  color: var(--prim-txt);
  padding: 0.91666667em 0.75em;
  vertical-align: inherit;
  font-style: none;
  font-weight: bold;
  text-transform: none;
  border-bottom: 1px solid var(--prim-brd);
  border-left: none;
}
.ui.table > thead > tr > th:first-child {
  border-left: none;
}
.ui.table > thead > tr:first-child > th:first-child {
  border-radius: 0.33333333rem 0 0 0;
}
.ui.table > thead > tr:first-child > th:last-child {
  border-radius: 0 0.33333333rem 0 0;
}
.ui.table > thead > tr:first-child > th:only-child {
  border-radius: 0.33333333rem 0.33333333rem 0 0;
}

/* Footer */
.ui.table > tfoot {
  -webkit-box-shadow: none;
          box-shadow: none;
}
.ui.table > tfoot > tr > th {
  cursor: auto;
  border-top: 1px solid var(--prim-brd);
  background: var(--sec-bg);
  text-align: inherit;
  color: var(--prim-txt);
  padding: 0.75em 0.75em;
  vertical-align: middle;
  font-style: normal;
  font-weight: normal;
  text-transform: none;
}
.ui.table > tfoot > tr > th:first-child {
  border-left: none;
}
.ui.table > tfoot > tr:first-child > th:first-child {
  border-radius: 0 0 0 0.33333333rem;
}
.ui.table > tfoot > tr:first-child > th:last-child {
  border-radius: 0 0 0.33333333rem 0;
}
.ui.table > tfoot > tr:first-child > th:only-child {
  border-radius: 0 0 0.33333333rem 0.33333333rem;
}

/* Table Row */
.ui.table > tr > td,
.ui.table > tbody > tr > td {
  border-top: 1px solid var(--prim-brd);
}
.ui.table > tr:first-child > td,
.ui.table > tbody > tr:first-child > td {
  border-top: none;
}

/* Repeated tbody */
.ui.table > tbody + tbody tr:first-child > td {
  border-top: 1px solid var(--prim-brd);
}

/* Table Cells */
.ui.table > tbody > tr > td,
.ui.table > tr > td {
  padding: 0.75em 0.75em;
  text-align: inherit;
}

/* Icons */
.ui.table > .icon {
  vertical-align: baseline;
}
.ui.table > .icon:only-child {
  margin: 0;
}

/* Table Segment */
.ui.table.segment {
  padding: 0;
}
.ui.table.segment:after {
  display: none;
}
.ui.table.segment.stacked:after {
  display: block;
}

/* Responsive */
@media only screen and (max-width: 767.98px) {
  .ui.table:not(.unstackable) {
    width: 100%;
    padding: 0;
  }
  .ui.table:not(.unstackable) > tbody,
  .ui.table:not(.unstackable) > tr,
  .ui.table:not(.unstackable) > tbody > tr,
  .ui.table:not(.unstackable) > tr > th,
  .ui.table:not(.unstackable) > thead > tr > th,
  .ui.table:not(.unstackable) > tbody > tr > th,
  .ui.table:not(.unstackable) > tfoot > tr > th,
  .ui.table:not(.unstackable) > tr > td,
  .ui.table:not(.unstackable) > tbody > tr > td {
    display: block !important;
    width: auto !important;
  }
  .ui.table:not(.unstackable) > thead {
    display: block;
  }
  .ui.table:not(.unstackable) > tfoot {
    display: block;
  }
  .ui.table:not(.unstackable) > tr,
  .ui.table:not(.unstackable) > thead > tr,
  .ui.table:not(.unstackable) > tbody > tr,
  .ui.table:not(.unstackable) > tfoot > tr {
    padding-top: 1em;
    padding-bottom: 1em;
    box-shadow: 0 -1px 0 0 rgba(0, 0, 0, 0.1) inset !important;
  }
  .ui.table:not(.unstackable) > tr > th,
  .ui.table:not(.unstackable) > thead > tr > th,
  .ui.table:not(.unstackable) > tbody > tr > th,
  .ui.table:not(.unstackable) > tfoot > tr > th,
  .ui.ui.ui.ui.table:not(.unstackable) > tr > td,
  .ui.ui.ui.ui.table:not(.unstackable) > tbody > tr > td {
    background: none;
    border: none;
    padding: 0.25em 0.75em;
    box-shadow: none !important;
  }
  .ui.table:not(.unstackable) > tr > th:first-child,
  .ui.table:not(.unstackable) > thead > tr > th:first-child,
  .ui.table:not(.unstackable) > tbody > tr > th:first-child,
  .ui.table:not(.unstackable) > tfoot > tr > th:first-child,
  .ui.table:not(.unstackable) > tr > td:first-child,
  .ui.table:not(.unstackable) > tbody > tr > td:first-child {
    font-weight: bold;
  }
  
/* Definition Table */
  .ui.definition.table:not(.unstackable) > thead > tr > th:first-child {
    box-shadow: none !important;
  }
}


/*******************************
            Coupling
*******************************/


/* UI Image */
.ui.table .collapsing .image,
.ui.table .collapsing .image img {
  max-width: none;
}


/*******************************
             Types
*******************************/


/*--------------
    Complex
---------------*/

.ui.structured.table {
  border-collapse: collapse;
}
.ui.structured.table > thead > tr > th {
  border-left: none;
  border-right: none;
}
.ui.structured.sortable.table > thead > tr > th {
  border-left: 1px solid var(--prim-brd);
  border-right: 1px solid var(--prim-brd);
}
.ui.structured.basic.table > tr > th,
.ui.structured.basic.table > thead > tr > th,
.ui.structured.basic.table > tbody > tr > th,
.ui.structured.basic.table > tfoot > tr > th {
  border-left: none;
  border-right: none;
}
.ui.structured.celled.table > tr > th,
.ui.structured.celled.table > thead > tr > th,
.ui.structured.celled.table > tbody > tr > th,
.ui.structured.celled.table > tfoot > tr > th,
.ui.structured.celled.table > tr > td,
.ui.structured.celled.table > tbody > tr > td {
  border-left: 1px solid var(--prim-brd);
  border-right: 1px solid var(--prim-brd);
}

/*--------------
   Definition
---------------*/

.ui.definition.table > thead:not(.full-width) > tr > th:first-child {
  pointer-events: none;
  background: transparent;
  font-weight: normal;
  color: var(--prim-txt);
  box-shadow: -1px -1px 0 1px #FFFFFF;
}
.ui.definition.table > tfoot:not(.full-width) > tr > th:first-child {
  pointer-events: none;
  background: transparent;
  font-weight: normal;
  color: var(--prim-txt);
  box-shadow: -1px 1px 0 1px #FFFFFF;
}

/* Remove Border */
.ui.celled.definition.table > thead:not(.full-width) > tr > th:first-child {
  box-shadow: 0 -1px 0 1px #FFFFFF;
}
.ui.celled.definition.table > tfoot:not(.full-width) > tr > th:first-child {
  box-shadow: 0 1px 0 1px #FFFFFF;
}

/* Highlight Defining Column */
.ui.definition.table > tr > td:first-child:not(.ignored),
.ui.definition.table > tbody > tr > td:first-child:not(.ignored),
.ui.definition.table tr td.definition {
  background: rgba(0, 0, 0, 0.03);
  font-weight: bold;
  color: #FFFFFF;
  text-transform: '';
  -webkit-box-shadow: '';
          box-shadow: '';
  text-align: '';
  font-size: 1em;
  padding-left: '';
  padding-right: '';
}

/* Fix 2nd Column */
.ui.definition.table > thead:not(.full-width) > tr > th:nth-child(2) {
  border-left: 1px solid var(--prim-brd);
}
.ui.definition.table > tfoot:not(.full-width) > tr > th:nth-child(2) {
  border-left: 1px solid var(--prim-brd);
}
.ui.definition.table > tr > td:nth-child(2),
.ui.definition.table > tbody > tr > td:nth-child(2) {
  border-left: 1px solid var(--prim-brd);
}


/*******************************
             States
*******************************/


/*--------------
    Positive
---------------*/

.ui.ui.ui.ui.table tr.positive,
.ui.ui.table td.positive {
  -webkit-box-shadow: 0 0 0 #A3C293 inset;
          box-shadow: 0 0 0 #A3C293 inset;
  background: #FCFFF5;
  color: #2C662D;
}

/*--------------
     Negative
---------------*/

.ui.ui.ui.ui.table tr.negative,
.ui.ui.table td.negative {
  -webkit-box-shadow: 0 0 0 #E0B4B4 inset;
          box-shadow: 0 0 0 #E0B4B4 inset;
  background: #B30015;
  color: #FFFFFF;
}

/*--------------
      Error
---------------*/

.ui.ui.ui.ui.table tr.error,
.ui.ui.table td.error {
  -webkit-box-shadow: 0 0 0 #E0B4B4 inset;
          box-shadow: 0 0 0 #E0B4B4 inset;
  background: #B30015;
  color: #FFFFFF;
}

/*--------------
     Warning
---------------*/

.ui.ui.ui.ui.table tr.warning,
.ui.ui.table td.warning {
  -webkit-box-shadow: 0 0 0 #C9BA9B inset;
          box-shadow: 0 0 0 #C9BA9B inset;
  background: #FFFAF3;
  color: #573A08;
}

/*--------------
     Active
---------------*/

.ui.ui.ui.ui.table tr.active,
.ui.ui.table td.active {
  box-shadow: 0 0 0 #B0C7D1 inset;
  background: var(--prim-button-bg);
  color: var(--prim-button-txt);
}

/*--------------
     Disabled
---------------*/

.ui.table tr.disabled td,
.ui.table tr td.disabled,
.ui.table tr.disabled:hover,
.ui.table tr:hover td.disabled {
  pointer-events: none;
  color: var(--prim-txt);
}


/*******************************
          Variations
*******************************/


/*--------------
    Stackable
---------------*/

@media only screen and (max-width: 991.98px) {
  .ui[class*="tablet stackable"].table,
  .ui[class*="tablet stackable"].table > tbody,
  .ui[class*="tablet stackable"].table > tbody > tr,
  .ui[class*="tablet stackable"].table > tr,
  .ui[class*="tablet stackable"].table > thead > tr > th,
  .ui[class*="tablet stackable"].table > tbody > tr > th,
  .ui[class*="tablet stackable"].table > tfoot > tr > th,
  .ui[class*="tablet stackable"].table > tr > th,
  .ui[class*="tablet stackable"].table > tbody > tr > td,
  .ui[class*="tablet stackable"].table > tr > td {
    display: block !important;
    width: 100% !important;
  }
  .ui[class*="tablet stackable"].table {
    padding: 0;
  }
  .ui[class*="tablet stackable"].table > thead {
    display: block;
  }
  .ui[class*="tablet stackable"].table > tfoot {
    display: block;
  }
  .ui[class*="tablet stackable"].table > thead > tr,
  .ui[class*="tablet stackable"].table > tbody > tr,
  .ui[class*="tablet stackable"].table > tfoot > tr,
  .ui[class*="tablet stackable"].table > tr {
    padding-top: 1em;
    padding-bottom: 1em;
    -webkit-box-shadow: 0 -1px 0 0 rgba(0, 0, 0, 0.1) inset !important;
            box-shadow: 0 -1px 0 0 rgba(0, 0, 0, 0.1) inset !important;
  }
  .ui[class*="tablet stackable"].table > thead > tr > th,
  .ui[class*="tablet stackable"].table > tbody > tr > th,
  .ui[class*="tablet stackable"].table > tfoot > tr > th,
  .ui[class*="tablet stackable"].table > tr > th,
  .ui[class*="tablet stackable"].table > tbody > tr > td,
  .ui[class*="tablet stackable"].table > tr > td {
    background: none;
    border: none !important;
    padding: 0.25em 0.75em;
    -webkit-box-shadow: none !important;
            box-shadow: none !important;
  }
  
/* Definition Table */
  .ui.definition[class*="tablet stackable"].table > thead > tr > th:first-child {
    -webkit-box-shadow: none !important;
            box-shadow: none !important;
  }
}

/*--------------
 Text Alignment
---------------*/

.ui.table[class*="left aligned"],
.ui.table [class*="left aligned"] {
  text-align: left;
}
.ui.table[class*="center aligned"],
.ui.table [class*="center aligned"] {
  text-align: center;
}
.ui.table[class*="right aligned"],
.ui.table [class*="right aligned"] {
  text-align: right;
}

/*------------------
 Vertical Alignment
------------------*/

.ui.table[class*="top aligned"],
.ui.table [class*="top aligned"] {
  vertical-align: top;
}
.ui.table[class*="middle aligned"],
.ui.table [class*="middle aligned"] {
  vertical-align: middle;
}
.ui.table[class*="bottom aligned"],
.ui.table [class*="bottom aligned"] {
  vertical-align: bottom;
}

/*--------------
    Collapsing
---------------*/

.ui.table th.collapsing,
.ui.table td.collapsing {
  width: 1px;
  white-space: nowrap;
}

/*--------------
     Fixed
---------------*/

.ui.fixed.table {
  table-layout: fixed;
}
.ui.fixed.table th,
.ui.fixed.table td {
  overflow: hidden;
  text-overflow: ellipsis;
}

/*--------------
   Selectable
---------------*/

.ui.ui.selectable.table > tbody > tr:hover,
.ui.table tbody tr td.selectable:hover {
  background: rgba(0, 0, 0, 0.05);
  color: #FFFFFF;
}
.ui.ui.selectable.inverted.table > tbody > tr:hover,
.ui.inverted.table tbody tr td.selectable:hover {
  background: rgba(255, 255, 255, 0.08);
  color: #000000;
}

/* Selectable Cell Link */
.ui.table tbody tr td.selectable {
  padding: 0;
}
.ui.table tbody tr td.selectable > a:not(.ui) {
  display: block;
  color: inherit;
  padding: 0.75em 0.75em;
}
.ui.table > tr > td.selectable,
.ui.table > tbody > tr > td.selectable,
.ui.selectable.table > tbody > tr,
.ui.selectable.table > tr {
  cursor: pointer;
}

/* Other States */
.ui.ui.selectable.table tr.error:hover,
.ui.table tr td.selectable.error:hover,
.ui.selectable.table tr:hover td.error {
  background: #a40013;
  color: #f7f7f7;
}
.ui.ui.selectable.table tr.warning:hover,
.ui.table tr td.selectable.warning:hover,
.ui.selectable.table tr:hover td.warning {
  background: #fff4e4;
  color: #493107;
}
.ui.ui.selectable.table tr.active:hover,
.ui.table tr td.selectable.active:hover,
.ui.selectable.table tr:hover td.active {
  background: #E0E0E0;
  color: var(--prim-txt);
}
.ui.ui.selectable.table tr.positive:hover,
.ui.table tr td.selectable.positive:hover,
.ui.selectable.table tr:hover td.positive {
  background: #f7ffe6;
  color: #275b28;
}
.ui.ui.selectable.table tr.negative:hover,
.ui.table tr td.selectable.negative:hover,
.ui.selectable.table tr:hover td.negative {
  background: #a40013;
  color: #f7f7f7;
}

/*-------------------
      Attached
--------------------*/


/* Middle */
.ui.attached.table {
  top: 0;
  bottom: 0;
  border-radius: 0;
  margin: 0 -1px;
  width: calc(100% + 2px);
  max-width: calc(100% + 2px);
  -webkit-box-shadow: none;
          box-shadow: none;
  border: 1px solid var(--prim-brd);
}
.ui.attached + .ui.attached.table:not(.top) {
  border-top: none;
}

/* Top */
.ui[class*="top attached"].table {
  bottom: 0;
  margin-bottom: 0;
  top: 0;
  margin-top: 1em;
  border-radius: 0.33333333rem 0.33333333rem 0 0;
}
.ui.table[class*="top attached"]:first-child {
  margin-top: 0;
}

/* Bottom */
.ui[class*="bottom attached"].table {
  bottom: 0;
  margin-top: 0;
  top: 0;
  margin-bottom: 1em;
  -webkit-box-shadow: none, none;
          box-shadow: none, none;
  border-radius: 0 0 0.33333333rem 0.33333333rem;
}
.ui[class*="bottom attached"].table:last-child {
  margin-bottom: 0;
}

/*--------------
     Striped
---------------*/


/* Table Striping */
.ui.striped.table > tr:nth-child(2n),
.ui.striped.table > tbody > tr:nth-child(2n) {
  background-color: rgba(0, 0, 50, 0.02);
}

/* Stripes */
.ui.inverted.striped.table > tr:nth-child(2n),
.ui.inverted.striped.table > tbody > tr:nth-child(2n) {
  background-color: rgba(255, 255, 255, 0.05);
}

/* Allow striped active hover */
.ui.striped.selectable.selectable.selectable.table tbody tr.active:hover {
  background: #EFEFEF;
  color: #FFFFFF;
}

/*--------------
   Single Line
---------------*/

.ui.table[class*="single line"],
.ui.table [class*="single line"] {
  white-space: nowrap;
}

/*-------------------
       Colors
--------------------*/

.ui.primary.table {
  border-top: 0.2em solid #8A1538;
}
.ui.inverted.primary.table {
  background-color: #8A1538;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.primary,
.ui.ui.table td.primary {
  -webkit-box-shadow: 0 0 0 #5e0e26 inset;
          box-shadow: 0 0 0 #5e0e26 inset;
  background: #e8d0d7;
  color: rgba(255, 255, 255, 0.9);
}
.ui.ui.selectable.table tr.primary:hover,
.ui.table tr td.selectable.primary:hover,
.ui.selectable.table tr:hover td.primary {
  background: #e0ced3;
  color: rgba(255, 255, 255, 0.9);
}
.ui.secondary.table {
  border-top: 0.2em solid #333F48;
}
.ui.inverted.secondary.table {
  background-color: #333F48;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.secondary,
.ui.ui.table td.secondary {
  -webkit-box-shadow: 0 0 0 #1e252a inset;
          box-shadow: 0 0 0 #1e252a inset;
  background: #d2d3d5;
  color: rgba(255, 255, 255, 0.9);
}
.ui.ui.selectable.table tr.secondary:hover,
.ui.table tr td.selectable.secondary:hover,
.ui.selectable.table tr:hover td.secondary {
  background: #d6d9db;
  color: rgba(255, 255, 255, 0.9);
}
.ui.red.table {
  border-top: 0.2em solid #8A1538;
}
.ui.inverted.red.table {
  background-color: #8A1538;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.red,
.ui.ui.table td.red {
  -webkit-box-shadow: 0 0 0 #5e0e26 inset;
          box-shadow: 0 0 0 #5e0e26 inset;
  background: #f0ccd0;
  color: #8A1538;
}
.ui.ui.selectable.table tr.red:hover,
.ui.table tr td.selectable.red:hover,
.ui.selectable.table tr:hover td.red {
  background: #e6cccf;
  color: #8A1538;
}
.ui.orange.table {
  border-top: 0.2em solid #ED8B00;
}
.ui.inverted.orange.table {
  background-color: #ED8B00;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.orange,
.ui.ui.table td.orange {
  -webkit-box-shadow: 0 0 0 #ba6d00 inset;
          box-shadow: 0 0 0 #ba6d00 inset;
  background: #ffe7d1;
  color: #ED8B00;
}
.ui.ui.selectable.table tr.orange:hover,
.ui.table tr td.selectable.orange:hover,
.ui.selectable.table tr:hover td.orange {
  background: #fae1cc;
  color: #ED8B00;
}
.ui.yellow.table {
  border-top: 0.2em solid #FBBD08;
}
.ui.inverted.yellow.table {
  background-color: #FBBD08;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.yellow,
.ui.ui.table td.yellow {
  -webkit-box-shadow: 0 0 0 #cd9903 inset;
          box-shadow: 0 0 0 #cd9903 inset;
  background: #fff9d2;
  color: #B58105;
}
.ui.ui.selectable.table tr.yellow:hover,
.ui.table tr td.selectable.yellow:hover,
.ui.selectable.table tr:hover td.yellow {
  background: #fbf5cc;
  color: #B58105;
}
.ui.olive.table {
  border-top: 0.2em solid #B5CC18;
}
.ui.inverted.olive.table {
  background-color: #B5CC18;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.olive,
.ui.ui.table td.olive {
  -webkit-box-shadow: 0 0 0 #8d9e13 inset;
          box-shadow: 0 0 0 #8d9e13 inset;
  background: #f7fae4;
  color: #8ABC1E;
}
.ui.ui.selectable.table tr.olive:hover,
.ui.table tr td.selectable.olive:hover,
.ui.selectable.table tr:hover td.olive {
  background: #f6fada;
  color: #8ABC1E;
}
.ui.green.table {
  border-top: 0.2em solid #B6BF00;
}
.ui.inverted.green.table {
  background-color: #B6BF00;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.green,
.ui.ui.table td.green {
  -webkit-box-shadow: 0 0 0 #858c00 inset;
          box-shadow: 0 0 0 #858c00 inset;
  background: #d5f5d9;
  color: #1EBC30;
}
.ui.ui.selectable.table tr.green:hover,
.ui.table tr td.selectable.green:hover,
.ui.selectable.table tr:hover td.green {
  background: #d2eed5;
  color: #1EBC30;
}
.ui.teal.table {
  border-top: 0.2em solid #00B5AD;
}
.ui.inverted.teal.table {
  background-color: #00B5AD;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.teal,
.ui.ui.table td.teal {
  -webkit-box-shadow: 0 0 0 #00827c inset;
          box-shadow: 0 0 0 #00827c inset;
  background: #e2ffff;
  color: #10A3A3;
}
.ui.ui.selectable.table tr.teal:hover,
.ui.table tr td.selectable.teal:hover,
.ui.selectable.table tr:hover td.teal {
  background: #d8ffff;
  color: #10A3A3;
}
.ui.blue.table {
  border-top: 0.2em solid #00A3E0;
}
.ui.inverted.blue.table {
  background-color: #00A3E0;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.blue,
.ui.ui.table td.blue {
  -webkit-box-shadow: 0 0 0 #007ead inset;
          box-shadow: 0 0 0 #007ead inset;
  background: #ddf4ff;
  color: #00A3E0;
}
.ui.ui.selectable.table tr.blue:hover,
.ui.table tr td.selectable.blue:hover,
.ui.selectable.table tr:hover td.blue {
  background: #d3f1ff;
  color: #00A3E0;
}
.ui.violet.table {
  border-top: 0.2em solid #6435C9;
}
.ui.inverted.violet.table {
  background-color: #6435C9;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.violet,
.ui.ui.table td.violet {
  -webkit-box-shadow: 0 0 0 #502aa1 inset;
          box-shadow: 0 0 0 #502aa1 inset;
  background: #ece9fe;
  color: #6435C9;
}
.ui.ui.selectable.table tr.violet:hover,
.ui.table tr td.selectable.violet:hover,
.ui.selectable.table tr:hover td.violet {
  background: #e3deff;
  color: #6435C9;
}
.ui.purple.table {
  border-top: 0.2em solid #A333C8;
}
.ui.inverted.purple.table {
  background-color: #A333C8;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.purple,
.ui.ui.table td.purple {
  -webkit-box-shadow: 0 0 0 #82299f inset;
          box-shadow: 0 0 0 #82299f inset;
  background: #f8e3ff;
  color: #A333C8;
}
.ui.ui.selectable.table tr.purple:hover,
.ui.table tr td.selectable.purple:hover,
.ui.selectable.table tr:hover td.purple {
  background: #f5d9ff;
  color: #A333C8;
}
.ui.pink.table {
  border-top: 0.2em solid #E03997;
}
.ui.inverted.pink.table {
  background-color: #E03997;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.pink,
.ui.ui.table td.pink {
  -webkit-box-shadow: 0 0 0 #c71f7e inset;
          box-shadow: 0 0 0 #c71f7e inset;
  background: #ffe8f9;
  color: #E03997;
}
.ui.ui.selectable.table tr.pink:hover,
.ui.table tr td.selectable.pink:hover,
.ui.selectable.table tr:hover td.pink {
  background: #ffdef6;
  color: #E03997;
}
.ui.brown.table {
  border-top: 0.2em solid #A5673F;
}
.ui.inverted.brown.table {
  background-color: #A5673F;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.brown,
.ui.ui.table td.brown {
  -webkit-box-shadow: 0 0 0 #805031 inset;
          box-shadow: 0 0 0 #805031 inset;
  background: #f7e5d2;
  color: #A5673F;
}
.ui.ui.selectable.table tr.brown:hover,
.ui.table tr td.selectable.brown:hover,
.ui.selectable.table tr:hover td.brown {
  background: #efe0cf;
  color: #A5673F;
}
.ui.grey.table {
  border-top: 0.2em solid #D7D9DB;
}
.ui.inverted.grey.table {
  background-color: #D7D9DB;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.grey,
.ui.ui.table td.grey {
  -webkit-box-shadow: 0 0 0 #bcc0c3 inset;
          box-shadow: 0 0 0 #bcc0c3 inset;
  background: #F0F1F2;
  color: #D7D9DB;
}
.ui.ui.selectable.table tr.grey:hover,
.ui.table tr td.selectable.grey:hover,
.ui.selectable.table tr:hover td.grey {
  background: #d4d8db;
  color: #D7D9DB;
}
.ui.black.table {
  border-top: 0.2em solid #1D252D;
}
.ui.inverted.black.table {
  background-color: #1D252D;
  color: #FFFFFF;
}
.ui.ui.ui.ui.table tr.black,
.ui.ui.table td.black {
  -webkit-box-shadow: 0 0 0 #090c0e inset;
          box-shadow: 0 0 0 #090c0e inset;
  background: #333F48;
  color: #FFFFFF;
}
.ui.ui.selectable.table tr.black:hover,
.ui.table tr td.selectable.black:hover,
.ui.selectable.table tr:hover td.black {
  background: #000000;
  color: #FFFFFF;
}

/*--------------
  Column Count
---------------*/


/* Grid Based */
.ui.one.column.table td {
  width: 100%;
}
.ui.two.column.table td {
  width: 50%;
}
.ui.three.column.table td {
  width: 33.33333333%;
}
.ui.four.column.table td {
  width: 25%;
}
.ui.five.column.table td {
  width: 20%;
}
.ui.six.column.table td {
  width: 16.66666667%;
}
.ui.seven.column.table td {
  width: 14.28571429%;
}
.ui.eight.column.table td {
  width: 12.5%;
}
.ui.nine.column.table td {
  width: 11.11111111%;
}
.ui.ten.column.table td {
  width: 10%;
}
.ui.eleven.column.table td {
  width: 9.09090909%;
}
.ui.twelve.column.table td {
  width: 8.33333333%;
}
.ui.thirteen.column.table td {
  width: 7.69230769%;
}
.ui.fourteen.column.table td {
  width: 7.14285714%;
}
.ui.fifteen.column.table td {
  width: 6.66666667%;
}
.ui.sixteen.column.table td {
  width: 6.25%;
}

/* Column Width */
.ui.table th.one.wide,
.ui.table td.one.wide {
  width: 6.25%;
}
.ui.table th.two.wide,
.ui.table td.two.wide {
  width: 12.5%;
}
.ui.table th.three.wide,
.ui.table td.three.wide {
  width: 18.75%;
}
.ui.table th.four.wide,
.ui.table td.four.wide {
  width: 25%;
}
.ui.table th.five.wide,
.ui.table td.five.wide {
  width: 31.25%;
}
.ui.table th.six.wide,
.ui.table td.six.wide {
  width: 37.5%;
}
.ui.table th.seven.wide,
.ui.table td.seven.wide {
  width: 43.75%;
}
.ui.table th.eight.wide,
.ui.table td.eight.wide {
  width: 50%;
}
.ui.table th.nine.wide,
.ui.table td.nine.wide {
  width: 56.25%;
}
.ui.table th.ten.wide,
.ui.table td.ten.wide {
  width: 62.5%;
}
.ui.table th.eleven.wide,
.ui.table td.eleven.wide {
  width: 68.75%;
}
.ui.table th.twelve.wide,
.ui.table td.twelve.wide {
  width: 75%;
}
.ui.table th.thirteen.wide,
.ui.table td.thirteen.wide {
  width: 81.25%;
}
.ui.table th.fourteen.wide,
.ui.table td.fourteen.wide {
  width: 87.5%;
}
.ui.table th.fifteen.wide,
.ui.table td.fifteen.wide {
  width: 93.75%;
}
.ui.table th.sixteen.wide,
.ui.table td.sixteen.wide {
  width: 100%;
}

/*--------------
    Sortable
---------------*/

.ui.sortable.table > thead > tr > th {
  cursor: pointer;
  white-space: nowrap;
  border-left: 1px solid var(--prim-brd);
  color: var(--prim-txt);
}
.ui.sortable.table > thead > tr > th:first-child {
  border-left: none;
}
.ui.sortable.table thead th.sorted,
.ui.sortable.table thead th.sorted:hover {
  -webkit-user-select: none;
     -moz-user-select: none;
      -ms-user-select: none;
          user-select: none;
}
.ui.sortable.table > thead > tr > th:after {
  display: none;
  font-style: normal;
  font-weight: normal;
  text-decoration: inherit;
  content: '';
  height: 1em;
  width: auto;
  opacity: 0.8;
  margin: 0 0 0 0.5em;
  font-family: 'Icons';
}
.ui.sortable.table thead th.ascending:after {
  content: '\f0d8';
}
.ui.sortable.table thead th.descending:after {
  content: '\f0d7';
}

/* Hover */
.ui.sortable.table th.disabled:hover {
  cursor: auto;
  color: var(--prim-txt);
}
.ui.sortable.table > thead > tr > th:hover {
  background: rgba(0, 0, 0, 0.05);
  color: var(--prim-txt);
}

/* Sorted */
.ui.sortable.table thead th.sorted {
  background: rgba(0, 0, 0, 0.05);
  color: #FFFFFF;
}
.ui.sortable.table thead th.sorted:after {
  display: inline-block;
}

/* Sorted Hover */
.ui.sortable.table thead th.sorted:hover {
  background: rgba(0, 0, 0, 0.05);
  color: #FFFFFF;
}

/* Inverted */
.ui.inverted.sortable.table thead th.sorted {
  background: rgba(255, 255, 255, 0.15) -webkit-gradient(linear, left top, left bottom, from(transparent), to(rgba(0, 0, 0, 0.05)));
  background: rgba(255, 255, 255, 0.15) -webkit-linear-gradient(transparent, rgba(0, 0, 0, 0.05));
  background: rgba(255, 255, 255, 0.15) linear-gradient(transparent, rgba(0, 0, 0, 0.05));
  color: #000000;
}
.ui.inverted.sortable.table > thead > tr > th:hover {
  background: rgba(255, 255, 255, 0.08) -webkit-gradient(linear, left top, left bottom, from(transparent), to(rgba(0, 0, 0, 0.05)));
  background: rgba(255, 255, 255, 0.08) -webkit-linear-gradient(transparent, rgba(0, 0, 0, 0.05));
  background: rgba(255, 255, 255, 0.08) linear-gradient(transparent, rgba(0, 0, 0, 0.05));
  color: #333F48;
}
.ui.inverted.sortable.table > thead > tr > th {
  border-left-color: transparent;
  border-right-color: transparent;
}

/*--------------
    Inverted
---------------*/


/* Text Color */
.ui.inverted.table {
  background: #333333;
  color: rgba(255, 255, 255, 0.9);
  border: none;
}
.ui.ui.inverted.table > thead > tr > th,
.ui.ui.inverted.table > tbody > tr > th,
.ui.ui.inverted.table > tfoot > tr > th,
.ui.ui.inverted.table > tr > th {
  background-color: rgba(0, 0, 0, 0.15);
  border-color: #1D252D;
  color: rgba(255, 255, 255, 0.9);
}
.ui.inverted.table > tbody > tr > td,
.ui.inverted.table > tr > td {
  border-color: #1D252D;
}
.ui.inverted.table tr.disabled td,
.ui.inverted.table tr td.disabled,
.ui.inverted.table tr.disabled:hover td,
.ui.inverted.table tr:hover td.disabled {
  pointer-events: none;
  color: #333F48;
}
.ui.inverted.table tr td.disabled:not([class="disabled"]),
.ui.inverted.table tr.disabled:not([class="disabled"]) td,
.ui.inverted.table tr.disabled td[class]:not(.disabled),
.ui.inverted.table tr:hover td.disabled:not([class="disabled"]) {
  color: var(--prim-txt);
}

/* Definition */
.ui.inverted.definition.table > tfoot:not(.full-width) > tr > th:first-child,
.ui.inverted.definition.table > thead:not(.full-width) > tr > th:first-child {
  color: var(--prim-bg);
}
.ui.inverted.definition.table > tbody > tr > td:first-child,
.ui.inverted.definition.table > tr > td:first-child {
  background: rgba(255, 255, 255, 0.02);
  color: #000000;
}

/*--------------
   Collapsing
---------------*/

.ui.collapsing.table {
  width: auto;
}

/*--------------
      Basic
---------------*/

.ui.basic.table {
  background: transparent;
  border: 1px solid var(--prim-brd);
  -webkit-box-shadow: none;
          box-shadow: none;
}
.ui.basic.table > thead,
.ui.basic.table > tfoot {
  -webkit-box-shadow: none;
          box-shadow: none;
}
.ui.basic.table > thead > tr > th,
.ui.basic.table > tbody > tr > th,
.ui.basic.table > tfoot > tr > th,
.ui.basic.table > tr > th {
  background: transparent;
  border-left: none;
}
.ui.basic.table > tbody > tr {
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}
.ui.basic.table > tbody > tr > td,
.ui.basic.table > tr > td {
  background: transparent;
}
.ui.basic.striped.table > tbody > tr:nth-child(2n) {
  background-color: rgba(0, 0, 0, 0.05);
}

/* Very Basic */
.ui[class*="very basic"].table {
  border: none;
}
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > th,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > thead > tr > th,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > th,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tfoot > tr > th,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > td,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > td {
  padding: '';
}
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > th:first-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > thead > tr > th:first-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > th:first-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tfoot > tr > th:first-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > td:first-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > td:first-child {
  padding-left: 0;
}
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > th:last-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > thead > tr > th:last-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > th:last-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tfoot > tr > th:last-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tr > td:last-child,
.ui[class*="very basic"].table:not(.sortable):not(.striped) > tbody > tr > td:last-child {
  padding-right: 0;
}
.ui[class*="very basic"].table:not(.sortable):not(.striped) > thead > tr:first-child > th {
  padding-top: 0;
}

/*--------------
     Celled
---------------*/

.ui.celled.table > tr > th,
.ui.celled.table > thead > tr > th,
.ui.celled.table > tbody > tr > th,
.ui.celled.table > tfoot > tr > th,
.ui.celled.table > tr > td,
.ui.celled.table > tbody > tr > td {
  border-left: 1px solid var(--prim-brd);
}
.ui.inverted.celled.table > tbody > tr > td,
.ui.inverted.celled.table > tr > td {
  border-left: 1px solid #1D252D;
}
.ui.celled.table > tr > th:first-child,
.ui.celled.table > thead > tr > th:first-child,
.ui.celled.table > tbody > tr > th:first-child,
.ui.celled.table > tfoot > tr > th:first-child,
.ui.celled.table > tr > td:first-child,
.ui.celled.table > tbody > tr > td:first-child {
  border-left: none;
}

/*--------------
     Padded
---------------*/

.ui.padded.table > tr > th,
.ui.padded.table > thead > tr > th,
.ui.padded.table > tbody > tr > th,
.ui.padded.table > tfoot > tr > th {
  padding-left: 1em;
  padding-right: 1em;
}
.ui.padded.table > tr > th,
.ui.padded.table > thead > tr > th,
.ui.padded.table > tbody > tr > th,
.ui.padded.table > tfoot > tr > th,
.ui.padded.table > tr > td,
.ui.padded.table > tbody > tr > td {
  padding: 1em 1em;
}

/* Very */
.ui[class*="very padded"].table > tr > th,
.ui[class*="very padded"].table > thead > tr > th,
.ui[class*="very padded"].table > tbody > tr > th,
.ui[class*="very padded"].table > tfoot > tr > th {
  padding-left: 1.5em;
  padding-right: 1.5em;
}
.ui[class*="very padded"].table > tr > td,
.ui[class*="very padded"].table > tbody > tr > td {
  padding: 1.5em 1.5em;
}

/*--------------
     Compact
---------------*/

.ui.compact.table > tr > th,
.ui.compact.table > thead > tr > th,
.ui.compact.table > tbody > tr > th,
.ui.compact.table > tfoot > tr > th {
  padding-left: 0.7em;
  padding-right: 0.7em;
}
.ui.compact.table > tr > td,
.ui.compact.table > tbody > tr > td {
  padding: 0.5em 0.7em;
}

/* Very */
.ui[class*="very compact"].table > tr > th,
.ui[class*="very compact"].table > thead > tr > th,
.ui[class*="very compact"].table > tbody > tr > th,
.ui[class*="very compact"].table > tfoot > tr > th {
  padding-left: 0.6em;
  padding-right: 0.6em;
}
.ui[class*="very compact"].table > tr > td,
.ui[class*="very compact"].table > tbody > tr > td {
  padding: 0.4em 0.6em;
}

/*--------------
      Sizes
---------------*/


/* Small */
.ui.small.table {
  font-size: 0.9em;
}

/* Standard */
.ui.table {
  font-size: 1em;
}

/* Large */
.ui.large.table {
  font-size: 1.1em;
}


/*******************************
         Site Overrides
*******************************/
`;
